# Coding Exercises (Appendix 3)

## Exercise 7: The Numpy Package: Working with numbers made easy!

1. Import Numpy!

In [1]:
import numpy as np

2. You have the following stock Prices in EUR Currency : [23.65, 43.21, 73.01, 87.32]<br>

Create a __Numpy array__ and __convert all prices__ to USD Currency @ an fx rate of __1.2__ (1.2 USD per 1 EUR).

In [2]:
rate = np.array([23.65, 43.21, 73.01, 87.32])

In [3]:
rate * 1.2

array([ 28.38 ,  51.852,  87.612, 104.784])

(result: array([ 28.38 ,  51.852,  87.612, 104.784])

You have the 
- end-of-2017 stock prices for four companies: [33.65, 63.21, 53.01, 17.32]
- end-of-2018 stock prices for the same companies: [36.54, 61.59, 60.31, 10.40]

3. Calculate the __absolute__ price increase / decrease in 2018 __for all four companies__. Use __vectorized__ numpy code!

In [4]:
p_2017 = np.array([33.65, 63.21, 53.01, 17.32])
p_2018 = np.array([36.54, 61.59, 60.31, 10.40])

In [5]:
p_2018 - p_2017

array([ 2.89, -1.62,  7.3 , -6.92])

4. Calculate the __price return__ in 2018 __for all four companies__. Use __vectorized__ numpy code!

In [6]:
(p_2018 / p_2017) -1 

array([ 0.0858841 , -0.02562886,  0.13770987, -0.39953811])

You have the following list with cashflows (starting today at t0): [-500, -100, 200, 600, 1000, 800]

In [7]:
#run the cell!
l  = [-500, -100, 200, 600, 1000, 800]

5. Convert l into a __numpy array__ and __save__ the array in memory! __Select__ the __second cashflow__ at t1.

In [8]:
a = np.array(l)
a

array([-500, -100,  200,  600, 1000,  800])

In [9]:
a[1]

-100

6. __Slice__ the array for the __last two cashflows__!

In [10]:
a[-2:]

array([1000,  800])

7. __Slice__ the array for the cashflows in __t2, t3 and t4__.

In [11]:
a[2:5]

array([ 200,  600, 1000])

The ABC Company evaluates to buy an additional machine that will increase future profits/cashflows by
- 50 USD in t1, 
- 80 USD in t2,
- 120 USD in t3, 
- 150 USD in t4,
- 200 USD in t5. (each cf at period´s end)<br>

8. The machine costs 500 USD (Investment in to). Calculate the __Project´s NPV__ and evaluate whether ABC should pursue the project. <br> ABC´s required rate of return (Cost of Capital) is 7.5% p.a. Use __vectorized__ numpy code!

In [12]:
# run the cell
cf = np.array([-500, 50, 80, 120, 150, 200])
cf

array([-500,   50,   80,  120,  150,  200])

In [13]:
r = 0.075
f = 1 + r

In [14]:
n = np.array([0, 1, 2, 3, 4, 5])
n

array([0, 1, 2, 3, 4, 5])

In [15]:
(cf / f**n).sum()

-36.03468883756554

(result: -36.03468883756554 -> no!)

9. __Slice__ the array cf (from question 8) for the positive inflows (__from t1__) and __save__ the slice in the __variable cf_pos__. <br> Make sure that once you __change the slice cf_pos__, the array __cf shall change as well__!

In [16]:
cf_pos = cf[1:]
cf_pos

array([ 50,  80, 120, 150, 200])

10. Change the first cashflow (t1) of __cf_pos to 70__. Check whether __cf gets mutated as well__!

In [17]:
cf_pos[0] = 70

In [18]:
cf

array([-500,   70,   80,  120,  150,  200])

11. __Restore cf__ and __redo Q9 and Q10__. This time make sure that cf and cf_pos are __independent objects__! 

In [19]:
cf = np.array([-500, 50, 80, 120, 150, 200])
cf

array([-500,   50,   80,  120,  150,  200])

In [20]:
cf_pos = cf[1:].copy()
cf_pos

array([ 50,  80, 120, 150, 200])

In [21]:
cf_pos[0] = 70
cf_pos

array([ 70,  80, 120, 150, 200])

In [22]:
cf

array([-500,   50,   80,  120,  150,  200])

12. Calculate the __total sum__ over all cashflows of cf!

In [23]:
# run the cell
cf = np.array([-500, 50, 80, 120, 150, 200])
cf

array([-500,   50,   80,  120,  150,  200])

In [24]:
cf.sum()

100

13. Return the __highest Value__ of cf!

In [25]:
cf.max()

200

14. Calculate the __cumulative cashflows__ for cf!

In [26]:
cf.cumsum()

array([-500, -450, -370, -250, -100,  100], dtype=int32)

15. Assume the following list: [3, 10, 6, 45, 100]. Calculate the __square root__ for all elements in the list. 

In [27]:
np.sqrt([3, 10, 6, 45, 100])

array([ 1.73205081,  3.16227766,  2.44948974,  6.70820393, 10.        ])

Assume the following stock prices:

In [28]:
# run the cell!
a = np.array([0.34, 43.56, 32.12, 67.43, 0.87, 9.87])
a

array([ 0.34, 43.56, 32.12, 67.43,  0.87,  9.87])

16. __Filter__ a for all penny stocks! (price less than 1)

In [29]:
a[a < 1]

array([0.34, 0.87])

(result: array([0.34, 0.87]))

# Well Done!